Modal prefix tag

nilsdeppe opened the pull request:
## Proposed changes

- Fixes a bug in DataBox that required closures be copyable when they don't actually need to be
- Add `ModalTag` to swap from a `tnsr` of nodal coefficients to a tensor of modal coefficients.

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add ModalTag prefix tag

nilsdeppe committed:
Fix type bug in DataBox

fmahebert commented on src/DataStructures/Tags.hpp:
```diff
@@ -0,0 +1,25 @@
+/// \endcond
+
+namespace Tags {
+/// Given the `NodalTag` swap the `Tensor::type` with a `ModalVector`.
```
I have a mild concern about the potential misuse of this tag. For example, if `NodalTag` wraps a `Tensor<double>` or a `Tensor<ComplexDataVector>`, then `Modal` will happily generate a valid tag ... but probably not the tag that was intended. In general we want our tags to validate types and avoid mistakes like this, so perhaps it makes sense to make this a little more "defensive".

I'm not sure if we should go to so far as to require from the compiler that `Tensor::type` be a `DataVector`, but perhaps the documentation should at least caution that the resulting Tag is (probably) only meaningful when `Tensor::type` is indeed a `DataVector`.

nilsdeppe replied:
I thought about this too. I think a `static_assert` here would be good.

fmahebert replied:
👍

nilsdeppe replied:
Pushed fixup :)

fmahebert replied:
Can you update this comment to reflect that the NodalTag is less generic now? I suggest "Wraps a `Tensor<ModalVector>` that has the same structure as the given `NodalTag`'s `Tensor<DataVector>`", or similar.

fmahebert reviewed (commented):
I don't feel qualified to review the Databox change, but here's a mild suggestion on the `Modal` Tag.

nilsvu commented on src/DataStructures/Tags.hpp:
```diff
@@ -0,0 +1,29 @@
+namespace Tags {
+/// Given the `NodalTag` swap the `Tensor::type` with a `ModalVector`.
+template <typename NodalTag>
+struct Modal : db::SimpleTag {
```
Inherit also from `PrefixTag`

nilsvu commented on src/DataStructures/Tags.hpp:
```diff
@@ -0,0 +1,29 @@
+/// Given the `NodalTag` swap the `Tensor::type` with a `ModalVector`.
+template <typename NodalTag>
+struct Modal : db::SimpleTag {
+  static_assert(cpp17::is_same_v<typename NodalTag::type::type, DataVector>,
```
Assert also that `db::item_type<NodalTag>` is a `Tensor`?

nilsdeppe replied:
The tensor type swap will fail to compile if that's not true, so I'd rather keep compile time down, personally.

nilsvu commented:
I, too, cannot sufficiently review the DataBox changes in the first commit. Is there a unit test covering that code? Can you add a test that would have caught the bug?

wthrowe commented:
Since the other reviewers are not confident in their ability to judge the DataBox changes, I'll jump in and state that they look correct to me.  I did not look at the rest of the PR.

nilsdeppe commented:
rebased on develop, added a test for DataBox change, addressed @nilsleiffischer inheritance comment.

fmahebert commented on src/DataStructures/Tags.hpp:
```diff
@@ -0,0 +1,30 @@
+struct Modal : db::SimpleTag, db::PrefixTag {
+  using tag = NodalTag;
+  static_assert(cpp17::is_same_v<typename NodalTag::type::type, DataVector>,
+                "The Modal tag should only be used to on tags that hold "
```
"used to on" -> "used on" ?

fmahebert reviewed (changes requested):
I am ready for you to squash previous fixups as well as these changes here.

nilsdeppe commented:
@fmahebert pushed fixup addressing your comments. I went for a slightly different comment, but hopefully you like it :) Once @nilsleiffischer is fine with the fixups I'll squash

fmahebert commented:
LGTM 👍

nilsvu reviewed (commented):
👍

nilsdeppe commented:
Squashed and rebased. Thanks for the feedback @nilsleiffischer and @fmahebert!!
